base/nullable.jl

```diff
@@ -217,3 +217,127 @@ function hash(x::Nullable, h::UInt)
         return hash(x.value, h + nullablehash_seed)
     end
 end
+
+"""
+    Lifted{F}
+
+A type used to represent the lifted version of a function `f::F`.
+
+Calling an `_f::Lifted{F}` on arguments `xs...` lowers to
+`lift(_f.f, U, xs...)`, where the return type parameter `U` is chosen with the
+help of type inference.
+"""
+immutable Lifted{F}
+    f::F
+    cache::Dict{Tuple{Vararg{DataType}}, DataType}
+
+    (::Type{Lifted}){F}(f::F) = new{F}(
+        f, Dict{Tuple{Vararg{DataType}}, DataType}()
+    )
+end
+
+function (_f::Lifted{F}){F}(xs...)
+    f, cache = _f.f, _f.cache
+    signature = map(eltype, xs)
+    U = Base.@get!(
+        cache,
+        signature,
+        Core.Inference.return_type(f, Tuple{signature...})
+    )
+    return lift(f, U, xs...)
+end
+
+"""
+    lift(f::F)::Lifted{F}
+
+Return a lifted version of `f`.
+"""
+lift(f) = Lifted(f)
+
+"""
+    lift(f, U, xs...)
+
+Return an empty `Nullable{U}` if any of the `xs` is null; otherwise, return the
+(`Nullable`-wrapped) value of `f` applied to the values of the `xs`.
+
+NOTE: There are two exceptions to the above: `lift(|, Bool, x, y)` and
+`lift(&, Bool, x, y)`. These methods both follow three-valued logic semantics.
+"""
+function lift(f, U::DataType, x)
+    if isnull(x)
+        return Nullable{U}()
+    else
+        return Nullable{U}(f(unsafe_get(x)))
```

Comment:
I use `unsafe_get` here because it is generic over `Nullable` and non-`Nullable` types, whereas `get(x)` for unqualified `x` is not defined.